xwiki-platform-core/xwiki-platform-resource/xwiki-platform-resource-default/src/main/java/org/xwiki/resource/internal/MainResourceReferenceHandlerManager.java

```diff
@@ -37,10 +41,19 @@
 @Singleton
 public class MainResourceReferenceHandlerManager extends AbstractResourceReferenceHandlerManager<ResourceType>
 {
+    @Inject
+    private ResourceReferenceHandlerManager<EntityResourceAction> entityResourceReferenceHandlerManager;
+
     @Override
-    protected boolean matches(ResourceReferenceHandler handler, ResourceType resourceType)
+    protected boolean matches(ResourceReferenceHandler handler, ResourceReference reference)
     {
-        return handler.getSupportedResourceReferences().contains(resourceType);
+        boolean handlerResourceSupported = handler.getSupportedResourceReferences()
+            .contains(extractResourceReferenceQualifier(reference));
+        if (handlerResourceSupported && handler instanceof EntityResourceReferenceHandler) {
```

Comment:
This is not good IMO. The MainResourceReferenceHandlerManager should not know about EntityResourceReferenceHandler

Reply:
I was expecting this comment, but I've no idea how to avoid that check if we want `#canHandle` to properly return false right away and not when `#handle` is called.

Reply:
yes it means we need to review the design and possibly change it.